README.md
# jwt-pwd &middot;  [![NPM](https://img.shields.io/npm/v/jwt-pwd.svg?style=flat)](https://www.npmjs.com/package/jwt-pwd) [![Tests](https://travis-ci.org/nicolasdao/jwt-pwd.svg?branch=master)](https://travis-ci.org/nicolasdao/jwt-pwd) [![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause) [![Neap](https://neap.co/img/made_by_neap.svg)](#this-is-what-we-re-up-to)
__*jwt-pwd*__ is a tiny crypto helper that helps building JWT (JSON Web Token, pronounced _jot_), hashing/salting and validating passwords using methods such as md5, sha1, sha256, sha512, ripemd160 and finally encrypt data using either AES or triple DES. It aims at facilitating the development of token based authentication and authorization APIs (e.g., REST, GraphQL). This package wraps the [jsonwebtoken package](https://www.npmjs.com/package/jsonwebtoken), the native NodeJS `crypto` package and the [node_hash package](https://www.npmjs.com/package/node_hash). 

# Table of Contents

> * [Install](#install) 
> * [Getting started](#getting-started) 
> 	- [Generating & validating JWTs](#generating--validating-jwts) 
> 	- [Hashing and salting password](#hashing-and-salting-password) 
>	- [Encrypting data](#encrypting-data)
>		- [AES (recommended)](#aes-recommended)
>		- [Triple DES](#triple-des)
> 	- [Authorizing HTTP Request With a JWT Token (Express)](#authorizing-http-request-with-a-jwt-token-express) 
> 	- [Other Utils](#other-utils) 
> * [FAQ](#faq) 
> 	- [How to generate a keys?](#how-to-generate-keys) 
>		- [Single key for symmetric algorithm](#single-key-for-symmetric-algorithm)
>		- [Private/public keys for asymmetric algorithms?](#privatepublic-keys-for-asymmetric-algorithms)
>			- [RSA key pair](#rsa-key-pair)
>			- [ECDSA key pair](#ecdsa-key-pair)
>	- [Why bearer tokens stored in cookies are not prefixed with bearer?](#why-bearer-tokens-stored-in-cookies-are-not-prefixed-with-bearer)
> * [About Neap](#this-is-what-we-re-up-to)
> * [License](#license)


# Install
```
npm i jwt-pwd
```

# Getting started
## Generating & validating JWTs

```js
const Crypto = require('jwt-pwd')
const { jwt } = new Crypto({ secret: 'your-jwt-secret' })
// Or you can also user
// const { jwt } = new Crypto()
// jwt.setKey('your-jwt-secret')

const claims = {
	id:1,
	email: 'you@coolcompany.com'
}

// 1. Create JWT
jwt.create(claims)
	.then(token => {
		console.log(`Your JWT: ${token}`)
		// 2. Validate JWT
		return jwt.validate(token) // validate returns a promise.
	})
	.then(validateClaims => {
		console.log(`User ID: ${validateClaims.id} - User email: ${validateClaims.email}`)
	})
	.catch(err => console.log(`Invalid token: ${err.message}`))

```

> WARNING: If the algorithm uses asymmetric keys, the public key has to be passed as follow to validate the token:
>	```js
>	jwt.validate(token, { key:publicKey })
>	```
> To learn more about using private/public keys, please refer to the example in the [Private/public keys for asymmetric algorithms?](#privatepublic-keys-for-asymmetric-algorithms) section.

To change the default algorithm, pass an option parameter as follow:

```js
jwt.create(claims, { algorithm:'HS512' })
```

The supported cryptographic algorithms are:
- `HS256` (default): HMAC signature with SHA-256 (symmetric key)
- `HS384`: HMAC signature with SHA-384 (symmetric key)
- `HS512`: HMAC signature with SHA-512 (symmetric key) 
- `RS256`: RSA signature with SHA-256 (asymmetric key)
- `RS384`: RSA signature with SHA-384 (asymmetric key) 
- `RS512`: RSA signature with SHA-512 (asymmetric key) 
- `PS256`: RSASSA-PSS signature with SHA-256 (asymmetric key) 
- `PS384`: RSASSA-PSS signature with SHA-384 (asymmetric key) 
- `PS512`: RSASSA-PSS signature with SHA-512 (asymmetric key) 
- `ES256`: ECDSA signature with SHA-256 (asymmetric key) 
- `ES384`: ECDSA signature with SHA-384 (asymmetric key) 
- `ES512`: ECDSA signature with SHA-512 (asymmetric key) 
- `none`: No signature

The key concept you must understand when it comes to choosing one of those algorithms is that they are mainly split in two categories:
- __Asymmetric algorithm__: The algorithm is using a private/public key to sign the token. The public key can be shared with the rest of the world so it can verify that the token has not been tampered. Because the private key is kept secret, there is very little risk of compromising the way the token's integrity .
- __Symmetric algorithm__: The algorithm is using a single key to sign the token. This means that if a third-party wishes to verify that the token has not been tampered, the two parties need to find a safe way to share the single key, which can adds complexity. 

Choose an asymmetric algorithm if you must let clients verifying the JWT without your intervention, othersise choose a symmetric algorithm as they are simpler to start with. Once you've choosen which type of algorithm fits your requirements, choosing a specific algorithm depends on the types of signature your ecosystem supports. If the JWT travels throughout multiple existing systems that must verify its integrity, then do some research on those systems to see what is the most secured common denominator between all those systems (i.e., the most secured asymmetric algorithm they all support), and then choose that one. If you are not constrained by third-party systems, and still need an asymmetric algorithm, _ES256_ is a good compromise between security and adoption. To learn more about generating keys, please refer to the [How to generate a keys?](#how-to-generate-keys) section.

## Hashing and salting password

> IMPORTANT: Hashing and salting is not the same as encrypting. Please refer to the Auth0 article [Adding Salt to Hashing: A Better Way to Store Passwords](https://auth0.com/blog/adding-salt-to-hashing-a-better-way-to-store-passwords/) to learn more. If you want to encrypt data, please jump to the [Encrypting data](#encrypting-data) section.

```js
const Crypto = require('jwt-pwd')
const { pwd } = new Crypto()

const password = 'your-super-safe-password'
const alg = 'sha512' // other options: md5, sha1, sha256, sha512, ripemd160

// 1. Hash and salt
const { salt, hashedSaltedPassword } = pwd.hashAndSalt({ password, alg })
console.log(`Encrypted password: ${hashedSaltedPassword} - Salt: ${salt}`)

// 2. Validate
console.log('Password validation result: ', pwd.validate({ password, hashedSaltedPassword, salt, alg })) 
console.log('Password validation result: ', pwd.validate({ password: '123', hashedSaltedPassword, salt, alg }))

```

## Encrypting data
### AES (recommended)

> The exact cipher is `aes-256-cbc`.

```js
const Crypto = require('jwt-pwd')
const { encryption } = new Crypto()

const data = { firstName:'Nic', secret:1234 }

// Randomly creates and sets the AES private key.
const encryptionKey = encryption.aes.setKey()
// Randomly creates and sets the initialization vector.
const initializationVector = encryption.aes.setIv()

console.log({
	encryptionKey,
	initializationVector
})

const { cipher, encrypted } = encryption.aes.encrypt(JSON.stringify(data))

console.log({ cipher, encrypted })

const decryptedData= JSON.parse(encryption.aes.decrypt(encrypted))

console.log(decryptedData)
```

Notice that the AES needs an `encryptionKey` and an `initializationVector` to function properly. Those variables must fits certain criteria based on the type of cipher used. 

The following snippet shows how to use your own key and iv:

```js
const Crypto = require('jwt-pwd')
const { encryption } = new Crypto()

encryption.aes.setKey(process.env.ENCRYPTION_KEY)
encryption.aes.setIv(process.env.IV)
```

### Triple DES

If you do not wish to use an initialization vector, you can use an older and less secure aldorithm called triple DES as follow:

```js
const Crypto = require('jwt-pwd')
const { encryption } = new Crypto()

const data = { firstName:'Nic', secret:1234 }

// Randomly creates and sets the DES private key.
const encryptionKey = encryption.des.setKey()

console.log({
	encryptionKey
})

const { cipher, encrypted } = encryption.des.encrypt(JSON.stringify(data))

console.log({ cipher, encrypted })

const decryptedData= JSON.parse(encryption.des.decrypt(encrypted))

console.log(decryptedData)
```

## Authorizing HTTP Request With a JWT Token (Express)

The following piece of code assume that a JWT token containing claims `{ firstName:'Nic' }` is passed to each request in the `Authorization` header. If the request is successfully authenticated, a new `claims` property is added to the `req` object. That property contains all the claims. If, on the contrary, the request fails the authentication handler, then a 403 code is immediately returned.

```js
const Crypto = require('jwt-pwd')
const { bearerHandler } = new Crypto({ jwtSecret: 'your-jwt-secret' })

app.get('/sayhi', bearerHandler(), (req,res) => res.status(200).send(`${req.claims.firstName} says hi.`))
```

#### bearerHandler(options)

* `options` `<Object>`
	- `key` 	`<String>` 	Default is 'Authorization'. This is the request's header that's supposed to contain the bearer token.
	- `cookie`  `<String>` 	Cookie's name storing the token. If specified, the token can be stored in that cookie.
	- `query`   `<String>` 	Query parameter's name storing the token. If specified, the token can be stored the request's query parameter (e.g., if query equals `code`, then the token can be passed using URL [https://neap.co/oauth2/token?code=32133213213123213](https://neap.co/oauth2/token?code=32133213213123213)).

## Other Utils
### Authorizing HTTP Request With an API Key (Express)

The following piece of code assume that an API key is passed in each request in the header `x-api-key` (this header key is configurable). If the request is successfully authenticated, the rest of the code is executed. If, on the contrary, the request fails the authentication handler, then a 403 code is immediately returned.

```js
const Crypto = require('jwt-pwd')
const { apiKeyHandler } = new Crypto({ jwtSecret: 'your-jwt-secret' })

app.get('/sayhello', apiKeyHandler({ key: 'x-api-key', value: 'your-api-key' }), (req,res) => res.status(200).send(`Hello`))
```

> NOTE: In this case, the `jwtSecret` is not involved in any encryption or validation. The `apiKeyHandler` is just a handy helper.

# FAQ
## How to generate keys?

The method to generate a keys or secrets depends on your business requirements. If you need to let third parties verify that JWTs have not been tampered, then you need an asymmetric algorithm so you can safely share the public key. If on the other hand signing your JWT is a one-way street, you can use a symmetric algorithm and generate a single secret.

### Single key for symmetric algorithm

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.2](https://github.com/nicolasdao/jwt-pwd/compare/v1.0.1...v1.0.2) (2020-09-23)

### [1.0.1](https://github.com/nicolasdao/jwt-pwd/compare/v1.0.0...v1.0.1) (2020-08-30)


### Bug Fixes

* Doc link is broken ([d7481cf](https://github.com/nicolasdao/jwt-pwd/commit/d7481cfbcaec40fec4981035566920f5ec2ac613))

## [1.0.0](https://github.com/nicolasdao/jwt-pwd/compare/v0.4.1...v1.0.0) (2020-08-29)


### Features

* Add support for encryption using AES or DES. Correct the doc which was containing misleading information regarding hashing, salting and encryption ([d5712e0](https://github.com/nicolasdao/jwt-pwd/commit/d5712e0d2d13d9dedaf7dbc94f6ffc0275a8824a))

### [0.4.1](https://github.com/nicolasdao/jwt-pwd/compare/v0.4.0...v0.4.1) (2020-08-28)

## [0.4.0](https://github.com/nicolasdao/jwt-pwd/compare/v0.3.6...v0.4.0) (2020-08-28)


### Bug Fixes

* Vulnerabilities ([857945a](https://github.com/nicolasdao/jwt-pwd/commit/857945acd14e8fabf2ef04d345913d8045a9e823))

### [0.3.6](https://github.com/nicolasdao/jwt-pwd/compare/v0.3.5...v0.3.6) (2020-07-16)

### [0.3.5](https://github.com/nicolasdao/jwt-pwd/compare/v0.3.4...v0.3.5) (2020-03-18)


### Bug Fixes

* Travis npm deploy ([48beafe](https://github.com/nicolasdao/jwt-pwd/commit/48beafe3b380d05687dcbd568cf5dc74b7c7d331))

### [0.3.4](https://github.com/nicolasdao/jwt-pwd/compare/v0.3.3...v0.3.4) (2020-03-18)


### Bug Fixes

* 17 of 21 vulnerabilities in 1306 scanned packages ([5f2b9a3](https://github.com/nicolasdao/jwt-pwd/commit/5f2b9a3edee0da077d53fb1eab266175b2073e9a))

### [0.3.3](https://github.com/nicolasdao/jwt-pwd/compare/v0.3.2...v0.3.3) (2019-07-22)


### Bug Fixes

* Vulnerability issues reported by npm and github ([7fd4919](https://github.com/nicolasdao/jwt-pwd/commit/7fd4919))



<a name="0.3.2"></a>
## [0.3.2](https://github.com/nicolasdao/jwt-pwd/compare/v0.3.1...v0.3.2) (2019-06-18)


### Bug Fixes

* Pach vulnerabilities ([712e1a6](https://github.com/nicolasdao/jwt-pwd/commit/712e1a6))



<a name="0.3.1"></a>
## [0.3.1](https://github.com/nicolasdao/jwt-pwd/compare/v0.3.0...v0.3.1) (2019-04-23)


### Features

* Add support for optional redirect url when the auth failed in the bearerHandler ([2dedceb](https://github.com/nicolasdao/jwt-pwd/commit/2dedceb))



<a name="0.3.0"></a>
# [0.3.0](https://github.com/nicolasdao/jwt-pwd/compare/v0.2.1...v0.3.0) (2019-04-22)


### Features

* Add support for cookie and query to transport tokens in the bearerHandler ([34fa297](https://github.com/nicolasdao/jwt-pwd/commit/34fa297))



<a name="0.2.1"></a>
## [0.2.1](https://github.com/nicolasdao/jwt-pwd/compare/v0.2.0...v0.2.1) (2019-04-20)



<a name="0.2.0"></a>
# [0.2.0](https://github.com/nicolasdao/jwt-pwd/compare/v0.1.2...v0.2.0) (2019-04-20)



<a name="0.1.2"></a>
## [0.1.2](https://github.com/nicolasdao/jwt-pwd/compare/v0.1.1...v0.1.2) (2019-03-18)


### Features

* Default the bearer token to the Authorization header ([1ff2240](https://github.com/nicolasdao/jwt-pwd/commit/1ff2240))



<a name="0.1.1"></a>
## [0.1.1](https://github.com/nicolasdao/jwt-pwd/compare/v0.1.0...v0.1.1) (2019-03-17)


### Bug Fixes

* JWT token with missing bearer token are valid ([3955d19](https://github.com/nicolasdao/jwt-pwd/commit/3955d19))



<a name="0.1.0"></a>
# [0.1.0](https://github.com/nicolasdao/jwt-pwd/compare/v0.0.3...v0.1.0) (2019-03-17)


### Bug Fixes

* Travis failed to deploy invalid option "--api_key=" ([de3d3fe](https://github.com/nicolasdao/jwt-pwd/commit/de3d3fe))


### Features

* Add support for Express handlers that authenticate based on JWT token or API key ([f9aa7f1](https://github.com/nicolasdao/jwt-pwd/commit/f9aa7f1))



<a name="0.0.3"></a>
## [0.0.3](https://github.com/nicolasdao/jwt-pwd/compare/v0.0.2...v0.0.3) (2019-03-16)



<a name="0.0.2"></a>
## 0.0.2 (2019-03-16)


### Features

* Add support for creating and validating JWT and encrypting and validating password ([ea0f9c3](https://github.com/nicolasdao/jwt-password-helper/commit/ea0f9c3))


index.js
/**
 * Copyright (c) 2017-2019, Neap Pty Ltd.
 * All rights reserved.
 * 
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
*/

/**
 * CRYPTO MANAGER
 * ------------------
 * 
 * Use the most common hashing algorithms ('md5', 'sha1', 'sha256', 'sha512', 'ripemd160') 
 * to hash data with a salt key. This is a typical way of managing 
 * password as well as validating them. The workflows are as follow:
 *
 * 		1. Storing a new password
 * 		-------------------------
 * 			a. Generate a new salt.
 * 			b. Hash the password with that salt.
 * 			c. Store both that hashed password and the salt in the DB.
 *    	
 *
 *		2. Validate the password
 *		------------------------
 *			a. Get the hashed password and the salt from the DB
 *			b. Hash the provided password with the salt
 *			c. Compare that re-hashed password with the stored hashed one. 
 */

const hash = require('node_hash')
const crypto = require('crypto')
const jwt = require('jsonwebtoken')
const url = require('url')

const hexToBuf = h => h ? Buffer.from(h, 'hex') : new Buffer(0)

/**
 * Creates new hex random salt.
 * 
 * @param  {Number} length Length of the random salt.
 * @return {String}        Salt
 */
const randomHexSalt = length => crypto.randomBytes(Math.ceil(length/2)).toString('hex') .slice(0,length)

/**
 * Hashes salted data using some of the common hashing algorithm.
 * 
 * @param  {String} data     	Data
 * @param  {String} alg   		'md5', 'sha1', 'sha256', 'sha512', 'ripemd160'
 * @param  {String} hexSalt 	Required Hexadecimal salt
 * @return {String}          	Hashed data
 */
const hashSaltedData = (data, alg, hexSalt) => {
	const saltBuf = hexToBuf(hexSalt)
	switch(alg) {
	case 'md5':
		return hash.md5(data, saltBuf)
	case 'sha1':
		return hash.sha1(data, saltBuf)
	case 'sha256':
		return hash.sha256(data, saltBuf)
	case 'sha512':
		return hash.sha512(data, saltBuf)
	case 'ripemd160':
		return hash.ripemd160(data, saltBuf)
	default:
		throw new Error(`Hash algorithm ${alg} is not supported.`)
	}
}


const validateData = (data, hashedData, alg, hexSalt) => {
	try {
		const reHashedData = hashSaltedData(data, alg, hexSalt)
		return hashedData == reHashedData
	}
	/*eslint-disable */
	catch(err) {
		/*eslint-enable */
		return false
	}
}

const getBearerFromCookie = (cname,req) => {
	const cookie = ((req || {}).headers || {}).cookie
	
	if (!cname || !cookie)
		return ''

	const name = cname + '='
	const decodedCookie = decodeURIComponent(cookie)
	const ca = decodedCookie.split(';')
	for(let i = 0; i <ca.length; i++) {
		let c = ca[i]
		while (c.charAt(0) == ' ')
			c = c.substring(1)
		if (c.indexOf(name) == 0) {
			const v = c.substring(name.length, c.length) 
			return v ? `Bearer ${v}` : ''
		}
	}
	return ''
}

const getBearerFromQuery = (qname,req) => {
	try {
		const u = (req||{}).url 
		if (!qname || !u)
			return ''

		const { search:querystring } = new url.URL(u)
		if (!querystring)
			return ''

		const obj = querystring.replace(/^\?/,'').split('&').reduce((acc,keyValue) => {
			if (keyValue) {
				const [key,...values] = keyValue.split('=')
				const value = decodeURIComponent(values.join('='))
				acc[key] = value
			}
			return acc
		}, {})
		const v = obj[qname] || ''
		return v ? `Bearer ${v}` : ''
	} catch(e) {
		return (() => '')(e)
	}
}

const ENCODING_OUTPUT_FORMAT = 'hex'
const TRIPLE_DES = 'des-ede3'
const AES = 'aes-256-cbc'
/**
 * Encrypts string using your own explicit algorithm or default back to AES ('aes-256-cbc') or Triple DES ('des-ede3') based on the
 * data available. 
 * 
 * @param  {String} text		
 * @param  {String} key					Encryption key.		
 * @param  {String} options.cipher		
 * @param  {String} options.iv			Initialization vector (required if 'optiions.cipher' is 'aes-256-cbc'
 * @param  {String} options.format		Default 'hex'. Valid values: 'base64', 'hex', 'buffer'
 * 
 * @return {String} output.encrypted
 * @return {String} output.iv
 */
const encrypt = (text, key, options) => {
	let { iv, cipher, format=ENCODING_OUTPUT_FORMAT } = options || {}
	const errorMsg = `Failed to encrypt data${cipher ? ` using the '${cipher}' cipher` : ''}`
	if (!key)
		throw new Error(`${errorMsg}. Missing required 'key'.`)

	if (format != 'hex' && format != 'base64' && format != 'buffer')
		throw new Error(`${errorMsg}. format '${format}' is not supported. Supported formats are: 'base64', 'hex', 'buffer'`)

	let encryptionKey = key
	if (!cipher) {
		if (!iv) {
			if (key.length < 24)
				throw new Error(`${errorMsg}. Invalid key length. To default to Triple DES cipher, the key must be 24 characters long.`)

			encryptionKey = key.slice(0,24)
			cipher = TRIPLE_DES
		} else {
			if (key.length < 32)
				throw new Error(`${errorMsg}. Invalid key length. To default to AES cipher, the key must be 32 characters long.`)
			
			encryptionKey = key.slice(0,32)
			cipher = AES
		}
	}

	if (cipher === AES) {
		if (!iv)
			throw new Error(`${errorMsg}. The ${AES} cipher requires an initialization vector ('iv' option).`)
		if (iv.length != 16)
			throw new Error(`${errorMsg}. The ${AES} requires an initialization vector ('iv' option) that is 16 characters long (current is ${iv.length})`)
	}

	try {
		const cipherObj = crypto.createCipheriv(cipher, encryptionKey, iv||null)
		const encryptedBuffer = Buffer.concat([cipherObj.update(text||''), cipherObj.final()])
		return {
			cipher,
			key: encryptionKey,
			encrypted: format == 'buffer' ? encryptedBuffer : encryptedBuffer.toString(format),
			iv:iv||null,
			format
		}
	} catch (err) {
		throw new Error(`${errorMsg}. ${err.message}`)
	}
}

/**
 * Decrypts string using the 'aes-256-cbc' cipher. 
 *
 * @param  {String} text	
 * @param  {String} key					Decryption key
 * @param  {String} iv					Initialization vector
 * @param  {String} options.cipher		Default 'aes-256-cbc'
 * @param  {String} options.iv			Initialization vector (required if 'optiions.cipher' is 'aes-256-cbc'
 * @param  {String} options.format		Default 'hex'. Valid values: 'base64', 'hex', 'buffer'
 * 
 * @return {String} decrypted
 */
const decrypt = (text, key, options) => {
	let { cipher, iv, format=ENCODING_OUTPUT_FORMAT } = options || {}
	cipher = cipher || AES
	const errorMsg = `Failed to decript data using the '${cipher}' cipher`

	if (text instanceof Buffer)
		format = null

	if (!key)
		throw new Error(`${errorMsg}. Missing required 'key'.`)

	if (cipher === AES) {
		if (!iv)
			throw new Error(`${errorMsg}. The ${AES} cipher requires an initialization vector ('iv' option).`)
		if (iv.length != 16)
			throw new Error(`${errorMsg}. The ${AES} requires an initialization vector ('iv' option) that is 16 characters long (current is ${iv.length})`)
	}

	try {
		const decipher = crypto.createDecipheriv(cipher, key, iv||null)
		const decrypted = Buffer.concat([decipher.update(text||'', format), decipher.final()]).toString()
		return decrypted
	} catch (err) {
		throw new Error(`${errorMsg}. ${err.message}`)
	}
}

/**
 * Merges the input secrets into the a consistant set of keys. 
 * 
 * @param  {Object} secret        Object or string. If it exists and all the others don't, then its value(s) are used to set all the others.
 * @param  {Object} jwtSecret     Object or string used to sign JWT
 * @param  {Object} encryptSecret [description]
 * @return {Object}                       [description]
 */
const _getSecrets = ({ secret, jwtSecret, encryptSecret }) => {
	let privateJwtKey, publicJwtKey, privateEncryptKey, publicEncryptKey, iv
	
	const secretType = secret ? typeof(secret) : null
	const jwtSecretType = jwtSecret ? typeof(jwtSecret) : null
	const encryptSecretType = encryptSecret ? typeof(encryptSecret) : null

	if (!jwtSecret && secret) {
		if (secretType == 'string') {
			privateJwtKey = secret
			publicJwtKey = secret
		} else if (secretType == 'object') {
			privateJwtKey = secret.privateKey
			publicJwtKey = secret.publicKey
		}
	} else if (jwtSecret) {
		if (jwtSecretType == 'string') {
			privateJwtKey = jwtSecret
			publicJwtKey = jwtSecret
		} else if (jwtSecretType == 'object') {
			privateJwtKey = jwtSecret.privateKey
			publicJwtKey = jwtSecret.publicKey
		}
	}

	if (!encryptSecret && secret) {
		if (secretType == 'string') {
			privateEncryptKey = secret
			publicEncryptKey = secret
		} else if (secretType == 'object') {
			privateEncryptKey = secret.privateKey
			publicEncryptKey = secret.publicKey
			iv = secret.iv
		}
	} else if (encryptSecret) {
		if (encryptSecretType == 'string') {
			privateEncryptKey = encryptSecret
			publicEncryptKey = encryptSecret
		} else if (encryptSecretType == 'object') {
			privateEncryptKey = encryptSecret.privateKey
			publicEncryptKey = encryptSecret.publicKey
			iv = encryptSecret.iv
		}
	}

	return { privateJwtKey, publicJwtKey, privateEncryptKey, publicEncryptKey, iv }
}

const generateAESkey = (format) => crypto.randomBytes(32).toString(format || 'hex').slice(0, 32)
const generateAESiv = (format) => crypto.randomBytes(16).toString(format || 'hex').slice(0, 16)
const generateDESkey = (format) => crypto.randomBytes(24).toString(format || 'hex').slice(0, 24)

/**
 * Creates a Crypto object. 
 * 
 * @param {String||Object} 	config.secret						key or object
 * @param {String} 			config.secret.privateKey		
 * @param {String} 			config.secret.publicKey		
 * @param {String} 			config.secret.iv					Initialization vector. Only used for encryption when the algorithm is AES
 * @param {String||Object} 	config.jwtSecret					key or object
 * @param {String} 			config.jwtSecret.privateKey		
 * @param {String} 			config.jwtSecret.publicKey		
 * @param {String||Object} 	config.encryptSecret				key or object
 * @param {String} 			config.encryptSecret.privateKey		
 * @param {String} 			config.encryptSecret.publicKey		
 * @param {String} 			config.encryptSecret.iv				Initialization vector. Only used for encryption when the algorithm is AES
 *
 * @return {Crypto}			crypto
 */
const Crypto = function(config) {
	const { secret, jwtSecret, encryptSecret } = config || {}
	let { privateJwtKey, publicJwtKey, privateEncryptKey, publicEncryptKey, iv } = _getSecrets({ secret, jwtSecret, encryptSecret })
	
	this.pwd = {
		/**
		 * Determines whether the password matches the hashed password in the DB
		 * 
		 * @param  {String} password          		Password sent from the client
		 * @param  {String} hashedSaltedPassword 	Password stored in the DB
		 * @param  {String} salt              		Salt stored in the DB
		 * @param  {String} alg  					Default is 'sha256'. Supported values: 'md5', 'sha1', 'sha256', 'sha512', 'ripemd160'
		 * @return {Boolean}                  
		 */
		validate: ({ password, hashedSaltedPassword, salt, alg='sha256' }) => validateData(password, hashedSaltedPassword, alg, salt),
		/**
		 * Hashes a salted password using the hash alg. 
		 * 
		 * @param  {String} password    				
		 * @param  {String} alg 						Default is 'sha256'. Supported values: 'md5', 'sha1', 'sha256', 'sha512', 'ripemd160'
		 * @return {String} output.salt 				
		 * @return {String} output.hashedSaltedPassword 	
		 */
		hashAndSalt: ({ password, alg='sha256' }) => {
			const salt = randomHexSalt(16)
			return {
				salt,
				hashedSaltedPassword: hashSaltedData(password, alg, salt) 
			}
		}
	}

	const _jwt = {
		setKey: (key, options) => {
			const { format, length } = options || {}
			privateJwtKey = key || crypto.randomBytes(length || 50).toString(format || 'base64')
			return privateJwtKey
		},
		getKey: () => privateJwtKey,
		/**
		 * Creates a JWT token
		 * 
		 * @param  {Object}  claims 			Optional, e.g., { id:1, email: 'nic@neap.co' }
		 * @param  {String}  options.algorithm	Default: 'RS256'. Supported: HS256, HS384, HS512, RS256, RS384, RS512, PS256, PS384, PS512, ES256, ES384, ES512
		 * @return {Promise}        			Promise resolving to a string.
		 */
		create: (claims={}, options) => new Promise((onSuccess, onFailure) => {
			if (!privateJwtKey)
				onFailure(new Error('Failed to create JWT. Missing required private key. Please use one of the following constructors to instantiate your Crypto object: new Crypto({ secret:\'priv-key\' }), new Crypto({ secret: { privateKey:\'priv-key\' } }),new Crypto({ jwtSecret:\'priv-key\' }), new Crypto({ jwtSecret: { privateKey:\'priv-key\' } }).'))
			jwt.sign(claims, privateJwtKey, options, (err, token) => {
				if (err)
					onFailure(err)
				else
					onSuccess(token)
			})
		}),

		/**
		 * Verifies whether or not the token is valid based on whether it can be decrypted or not. 
		 * 
		 * @param  {String} 	token 
		 * @param  {String}     options.key			Public key to decrypt the JWT.
		 * @param  {[String]}   options.algorithms 
		 * @return {Promise}    Promise resolving to a Claims object
		 */
		validate: (token='', options) => new Promise((onSuccess, onFailure) => {
			const { key, algorithms } = options || {}
			const cert = key || publicJwtKey || privateJwtKey
			const opts = algorithms ? { algorithms } : {}

			if (!cert)
				onFailure(new Error('Failed to verify JWT. Missing required \'key\'/'))

			jwt.verify(token, cert, opts, (err, claims) => {
				if (err)
					onFailure(err)
				else
					onSuccess(claims)
			})
		}),

		/**
		 * Decodes a JWT. 
		 * 
		 * @param  {String}  token
		 * @param  {Boolean} options.complete	Default false. When set to true, the decoding includes the header and the signature. 		
		 * @return {Object}
		 */
		decode: (token, options) => jwt.decode(token, options)
	}

	this.encryption = {
		aes: {
			setKey: key => {
				privateEncryptKey = key || generateAESkey()
				return privateEncryptKey
			},
			getKey: () => privateEncryptKey,
			setIv: () => {
				iv = generateAESiv()
				return iv
			},
			generateKey: generateAESkey,
			generateIv: generateAESiv,
			/**
			 * Encrypts string using the 'aes-256-cbc' cipher. 
			 * 
			 * @param  {String} text				
			 * @param  {String} options.format		Default 'hex'. Valid values: 'base64', 'hex', 'buffer'
			 * 
			 * @return {String} output.encrypted
			 * @return {String} output.iv
			 */
			encrypt: (text, options={}) => encrypt(text, privateEncryptKey, { ...options, iv, cipher:AES }),
			/**
			 * Decrypts string using the 'aes-256-cbc' cipher. 
			 *
			 * @param  {String} text	
			 * @param  {String} options.key			
			 * @param  {String} options.format		Default 'hex'. Valid values: 'base64', 'hex', 'buffer'
			 * 
			 * @return {String} decrypted
			 */
			decrypt: (text, options={}) => decrypt(text, options.key || publicEncryptKey || privateEncryptKey, { ...options, iv, cipher:AES })
		},
		des: {
			setKey: key => {
				privateEncryptKey = key || generateDESkey()
				return privateEncryptKey
			},
			generateKey: generateDESkey,
			/**
			 * Encrypts string using the 'aes-256-cbc' cipher. 
			 * 
			 * @param  {String} text				
			 * @param  {String} options.format		Default 'hex'. Valid values: 'base64', 'hex', 'buffer'
			 * 
			 * @return {String} output.encrypted
			 * @return {String} output.iv
			 */
			encrypt: (text, options={}) => encrypt(text, privateEncryptKey, { ...options, iv, cipher:TRIPLE_DES }),
			/**
			 * Decrypts string using the 'aes-256-cbc' cipher. 
			 *
			 * @param  {String} text	
			 * @param  {String} options.key			
			 * @param  {String} options.format		Default 'hex'. Valid values: 'base64', 'hex', 'buffer'
			 * 
			 * @return {String} decrypted
			 */
			decrypt: (text, options={}) => decrypt(text, options.key || publicEncryptKey || privateEncryptKey, { ...options, iv, cipher:TRIPLE_DES })
		},
		/**
		 * Encrypts string using the most appropriate cipher. 
		 * 
		 * @param  {String} text				
		 * @param  {String} options.cipher		
		 * @param  {String} options.format		Default 'hex'. Valid values: 'base64', 'hex', 'buffer'
		 * 
		 * @return {String} output.encrypted
		 * @return {String} output.iv
		 */
		encrypt: (text, options={}) => encrypt(text, privateEncryptKey, { ...options, iv }),
		/**
		 * Decrypts string using the most appropriate cipher. 
		 *
		 * @param  {String} text	
		 * @param  {String} options.cipher		Default 'aes-256-cbc'
		 * @param  {String} options.key			
		 * @param  {String} options.format		Default 'hex'. Valid values: 'base64', 'hex', 'buffer'
		 * 
		 * @return {String} decrypted
		 */
		decrypt: (text, options={}) => decrypt(text, options.key || publicEncryptKey || privateEncryptKey, { ...options, iv })
	}

	this.jwt = _jwt

	this.apiKeyHandler = ({ key, value }) => {
		if (!key)
			throw new Error('Missing required argument \'key\'')
		if (!value)
			throw new Error('Missing required argument \'value\'')

		return (req,res,next) => {
			const keyValue = (req.headers || {})[key]
			if (keyValue != value)
				res.status(403).send(`Unauthorized access. ${keyValue ? `Invalid API key '${key}'.` : `Missing API key. Header '${key}' not found.`}`)
			next()
		}
	}

	this.bearerHandler = (options) => {
		const { key='Authorization', cookie, query, redirectUrl } = options || {}
		return (req,res,next) => {
			const headers = req.headers || {}
			const keyValue = key == 'Authorization' ? (headers[key] || headers['authorization']) : headers[key]
			const cookieValue = cookie ? getBearerFromCookie(cookie, req) : ''
			const queryValue = query ? getBearerFromQuery(query, req) : ''
			const bearerToken = cookieValue || queryValue ||keyValue
			if (!bearerToken) {
				redirectUrl 
					? res.redirect(redirectUrl)
					: res.status(403).send(`Unauthorized access. Missing bearer token. Header '${key}' not found.`)
				next()
			} else if (!/^[bB]earer\s/.test(bearerToken)) {
				redirectUrl 
					? res.redirect(redirectUrl)
					: res.status(403).send('Unauthorized access. Malformed bearer token. Missing bearer schema.')
				next()
			} else {
				const token = bearerToken.trim().replace(/^[bB]earer\s/, '')
				_jwt.validate(token)
					.catch(err => redirectUrl 
						? res.redirect(redirectUrl)
						: res.status(403).send(`Unauthorized access. Invalid bearer token. ${err.message}`))
					.then(claims => {
						if (claims)
							req.claims = claims 
						next()
					})
			}
		}
	}

	return this
}

module.exports = Crypto






